### PR TITLE
A new implementation for assigns clauses

### PR DESCRIPTION
**Pull request by feliperodri**
This change is a complete rewrite of the assigns clause, based on a new formalization. Each assigns clause target is modeled as a "conditional address range" (CAR) that is an address range guarded by a validity condition.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a ~~My commit message includes data points confirming performance improvements (if claimed).~~
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by jimgrundy — APPROVED** [review]

**Review by jimgrundy — APPROVED** [review]

File: src/goto-instrument/contracts/utils.h [hunk 1]
```diff
@@ -81,12 +81,25 @@ exprt generate_lexicographic_less_than_check(
...
   goto_programt::targett &target,
   goto_programt &payload);
 
+/// \brief Adds a new temporary symbol to the symbol table.
+///
+/// \param type: The type of the new symbol.
+/// \param location: The source location for the new symbol.
+/// \param mode: The mode for the new symbol, e.g. ID_C, ID_java.
+/// \param symtab: The symbol table to which the new symbol is to be added.
+/// \return The new symbolt object.
```
Comment (feliperodri) [inline src/goto-instrument/contracts/utils.h:98]: Nice!

File: src/ansi-c/ansi_c_convert_type.cpp [hunk 2]
```diff
@@ -14,6 +14,7 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <util/c_types.h>
 #include <util/config.h>
 #include <util/expr_util.h>
+#include <util/std_code.h>
```
Comment (feliperodri) [inline src/ansi-c/ansi_c_convert_type.cpp]: If we don't use `has_subexpr` in this file anymore, you can also remove the `expr_util.h` include.

**Review by jimgrundy — APPROVED** [review]

File: regression/contracts/assigns_enforce_23/main.c [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+void foo(blob *b, uint8_t *value)
+  // clang-format off
+__CPROVER_requires(b->size == 5)
+#if 0
+__CPROVER_assigns(__CPROVER_object(b->buf))
+__CPROVER_assigns(__CPROVER_object(value))
+#else
+__CPROVER_assigns(*(b->buf))
+__CPROVER_assigns(*(value))
+#endif
```
Comment (feliperodri) [inline regression/contracts/assigns_enforce_23/main.c]: The first part is dead code. We should update it.
```c
__CPROVER_assigns(__CPROVER_object(b->buf))
__CPROVER_assigns(__CPROVER_object(value))
```

File: regression/contracts/assigns_enforce_23/test.desc [hunk 4]
```diff
@@ -0,0 +1,13 @@
+KNOWNBUG
```
Comment (feliperodri) [inline regression/contracts/assigns_enforce_23/test.desc]: Why is this a known bug?

File: regression/contracts/assigns_enforce_23/test.desc [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+--
+--
+This test checks that verification succeeds if only expressions inside the assigns clause are assigned within the function.
+
+Note: For all 'enforce' tests, nothing can be assumed about the return value of the function (as the function call is not replaced at this point).
+
+To make such assumptions would cause verification to fail.
```
Comment (feliperodri) [inline regression/contracts/assigns_enforce_23/test.desc]: This comment seems unrelated to this test, could you update it?

File: regression/contracts/assigns_enforce_elvis_4/test.desc [hunk 6]
```diff
@@ -1,9 +1,9 @@
-KNOWNBUG
+CORE
 main.c
 --enforce-contract foo
 ^EXIT=1$
 ^SIGNAL=0$
-^.*error: illegal target in assigns clause$
+^.*error: void-typed targets not permitted$
```
Comment (feliperodri) [inline regression/contracts/assigns_enforce_elvis_4/test.desc]: I updated this on https://github.com/diffblue/cbmc/pull/6371. Could you rebase again?

File: regression/contracts/assigns_enforce_elvis_5/test.desc [hunk 7]
```diff
@@ -8,3 +8,5 @@ main.c
 --
 Check that Elvis operator expressions of form '(cond ? *if_true : *if_false)'
 are supported in assigns clauses.
+
+BUG: `is_lvalue` in goto is not consistent with `target.get_bool(ID_C_lvalue)`.
```
Comment (feliperodri) [inline regression/contracts/assigns_enforce_elvis_5/test.desc:12]: We agreed that CBMC is doing the reasonable thing here. I also updated it on the previous PR to `CORE` instead of `KNOWNBUG`.

File: regression/contracts/assigns_enforce_function_calls/test.desc [hunk 8]
```diff
@@ -1,9 +1,9 @@
 CORE
 main.c
 --enforce-contract foo
-^EXIT=(1|64)$
+^EXIT=1$
 ^SIGNAL=0$
-^.*error: Assigns clause is not side-effect free.$
+^.*Only calls to __CPROVER_object are allowed in assigns clauses\.$
```
Comment (feliperodri) [inline regression/contracts/assigns_enforce_function_calls/test.desc]: Nice!

File: regression/contracts/assigns_enforce_object_wrong_args/test.desc [hunk 9]
```diff
@@ -0,0 +1,10 @@
+CORE
+main.c
+--enforce-contract baz
+^EXIT=1$
+^SIGNAL=0$
+^.*__CPROVER_object must be invoked with a single argument\.$
```
Comment (feliperodri) [inline regression/contracts/assigns_enforce_object_wrong_args/test.desc]: Nice!

**Comment by SaswatPadhi** [comment]
Depends on #6371 -- must be merged after that is merged.

File: src/goto-instrument/contracts/assigns.h [hunk 10]
```diff
@@ -14,61 +14,90 @@ Date: July 2021
 #ifndef CPROVER_GOTO_INSTRUMENT_CONTRACTS_ASSIGNS_H
 #define CPROVER_GOTO_INSTRUMENT_CONTRACTS_ASSIGNS_H
 
+#include <iostream>
```
Comment (feliperodri) [inline src/goto-instrument/contracts/assigns.h]: Remove debug code.

File: src/goto-instrument/contracts/assigns.h [hunk 11]
```diff
@@ -14,61 +14,90 @@ Date: July 2021
...
-  /// \brief A class for representing targets for assigns clauses
-  class targett
+  /// \brief A class for representing Conditional Address Ranges
+  class conditional_address_ranget
   {
   public:
-    targett(const assigns_clauset &, const exprt &);
+    conditional_address_ranget(const assigns_clauset &, const exprt &);
+
+    exprt generate_inclusion_check(
```
Comment (feliperodri) [inline src/goto-instrument/contracts/assigns.h]: Could you add a comment header to it?

File: src/goto-instrument/contracts/assigns.h [hunk 12]
```diff
@@ -14,61 +14,90 @@ Date: July 2021
...
+      const conditional_address_ranget &lhs,
+      bool skip_lhs_validity_check = false) const;
 
-    static exprt normalize(const exprt &);
+    goto_programt generate_snapshot_instructions() const;
```
Comment (feliperodri) [inline src/goto-instrument/contracts/assigns.h:35]: Could you add a comment header to it? Same for the remaining methods.

File: src/goto-instrument/contracts/utils.h [hunk 13]
```diff
@@ -28,6 +29,11 @@ class havoc_if_validt : public havoc_utilst
   {
   }
 
+  void append_havoc_code_for_expr(
+    const source_locationt location,
+    const exprt &expr,
+    goto_programt &dest) const override;
```
Comment (feliperodri) [inline src/goto-instrument/contracts/utils.h]: Could you add a comment header to it?

File: src/goto-instrument/contracts/contracts.cpp [hunk 14]
```diff
@@ -779,28 +810,40 @@ bool code_contractst::check_frame_conditions_function(const irep_idt &function)
...
     {
-      // Local variables are always freely assignable
-      assigns.add_to_write_set(instruction_it->get_decl().symbol());
+      const auto car =
+        assigns.add_to_write_set(instruction_it->get_decl().symbol());
+      auto snapshot_instructions = car->generate_snapshot_instructions();
+      instruction_it++;
+      insert_before_swap_and_advance(
+        body, instruction_it, snapshot_instructions);
+      instruction_it--;
```
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.cpp]: Why do we need to modify `instruction_it` like this here?

File: src/goto-instrument/contracts/contracts.cpp [hunk 15]
```diff
@@ -665,24 +679,37 @@ void code_contractst::instrument_call_statement(
...
   else if(called_function == "free")
   {
-    const exprt &lhs_dereference = dereference_exprt(
-      to_typecast_expr(instruction_it->call_arguments().front()).op());
-    add_containment_check(body, assigns, instruction_it, lhs_dereference);
-    assigns.remove_from_write_set(lhs_dereference);
+    const auto &object = side_effect_expr_function_callt{
+      symbol_exprt::typeless(CPROVER_PREFIX "object"),
+      {instruction_it->call_arguments().front()},
+      /* the type does not matter -- we would destruct the call immediately */
```
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.cpp]: Let's keep comment style `//`

File: src/goto-instrument/contracts/contracts.cpp [hunk 16]
```diff
@@ -665,24 +679,37 @@ void code_contractst::instrument_call_statement(
...
-      const exprt &lhs = instruction_it->assign_lhs();
-      assigns.add_to_write_set(dereference_exprt(lhs));
+      const auto &object = function_call.lhs();
+
+      instruction_it++;
+
+      const auto car = assigns.add_to_write_set(side_effect_expr_function_callt{
+        symbol_exprt::typeless(CPROVER_PREFIX "object"),
+        {object},
+        /* the type does not matter -- we would destruct the call immediately */
```
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.cpp]: Let's keep comment style `//`

File: src/goto-instrument/contracts/contracts.cpp [hunk 17]
```diff
@@ -665,24 +679,37 @@ void code_contractst::instrument_call_statement(
...
+        empty_typet{},
+        instruction_it->source_location});
+      auto snapshot_instructions = car->generate_snapshot_instructions();
+      insert_before_swap_and_advance(
+        body, instruction_it, snapshot_instructions);
+
+      instruction_it--;
```
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.cpp:675]: If the next assignment is `return_malloc = ...` we would annotate this assignment with an assigns clause in the next iteration, right? Shall we jump the next instructions if that's the case?

File: src/goto-instrument/contracts/contracts.cpp [hunk 18]
```diff
@@ -665,24 +679,37 @@ void code_contractst::instrument_call_statement(
...
+    const auto &function_call =
+      to_code_function_call(instruction_it->get_code());
+    if(function_call.lhs().is_not_nil())
     {
-      instrument_assign_statement(instruction_it, body, assigns);
-      const exprt &lhs = instruction_it->assign_lhs();
-      assigns.add_to_write_set(dereference_exprt(lhs));
+      const auto &object = function_call.lhs();
+
+      instruction_it++;
```
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.cpp:667]: Why do we need to iterate here? Could you add a comment?

File: src/goto-instrument/contracts/contracts.cpp [hunk 19]
```diff
@@ -643,7 +650,14 @@ void code_contractst::instrument_assign_statement(
       lhs_sym.is_static_lifetime &&
       lhs_sym.location.get_function() ==
         instruction_iterator->source_location.get_function())
-      assigns_clause.add_to_write_set(lhs);
+    {
+      // TODO: Fix this.
+      // The CAR snapshot should be made only once at the beginning.
```
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.cpp]: This comment should map to a GitHub issue and we should add the link here
```c
TODO: <link>
```

File: src/goto-instrument/contracts/assigns.cpp [hunk 20]
```diff
@@ -9,144 +9,179 @@ Date: July 2021
...
+     same_object(lhs.upper_bound_address_var, upper_bound_address_var),
+     less_than_or_equal_exprt{lower_bound_address_var,
+                              lhs.lower_bound_address_var},
+     less_than_or_equal_exprt{lhs.upper_bound_address_var,
+                              upper_bound_address_var}});
+
+  if(skip_lhs_validity_check)
+    return inclusion_check;
+
+  return or_exprt{not_exprt{lhs.validity_condition_var}, inclusion_check};
```
Comment (feliperodri) [inline src/goto-instrument/contracts/assigns.cpp]: Is this suppose to be unreachable?

File: src/goto-instrument/contracts/assigns.cpp [hunk 21]
```diff
@@ -9,144 +9,179 @@ Date: July 2021
...
 }
 
-exprt assigns_clauset::targett::normalize(const exprt &expr)
+const slicet assigns_clauset::conditional_address_ranget::normalize_to_slice(
+  const exprt &expr) const
 {
-  if(expr.id() != ID_address_of)
-    return expr;
+  // FIXME: Refactor these checks out to a common function that can be
+  // used both in compilation and instrumentation stages
```
Comment (feliperodri) [inline src/goto-instrument/contracts/assigns.cpp]: This should map to a GitHub issue, and we should add the link here.

File: src/goto-instrument/contracts/contracts.cpp (hunk 14 above)
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/contracts.cpp]: Added a comment here.

File: src/goto-instrument/contracts/contracts.cpp (hunk 17 above)
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/contracts.cpp:675]: Added a comment here.

File: src/goto-instrument/contracts/contracts.cpp (hunk 18 above)
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/contracts.cpp:667]: Added a comment here.

File: src/goto-instrument/contracts/assigns.cpp (hunk 20 above)
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/assigns.cpp]: Yes, removed dead code and made it private.

File: src/goto-instrument/contracts/assigns.h (hunk 11 above)
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/assigns.h]: I added some comments within the functions, but since are short on time, I think we should defer doxygen comments on all of the contracts functions to a separate PR.

File: src/goto-instrument/contracts/assigns.h (hunk 12 above)
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/assigns.h:35]: I added some comments within the functions, but since are short on time, I think we should defer doxygen comments on all of the contracts functions to a separate PR.

File: src/goto-instrument/contracts/utils.h (hunk 13 above)
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/utils.h]: I added some comments within the functions, but since are short on time, I think we should defer doxygen comments on all of the contracts functions to a separate PR.

File: src/ansi-c/ansi_c_convert_type.cpp [hunk 22]
```diff
@@ -13,7 +13,7 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include <util/c_types.h>
 #include <util/config.h>
-#include <util/expr_util.h>
+#include <util/std_code.h>
```
Comment (feliperodri) [inline src/ansi-c/ansi_c_convert_type.cpp:16]: Why do we need this include?

File: src/goto-instrument/contracts/assigns.cpp [hunk 23]
```diff
@@ -9,144 +9,177 @@ Date: July 2021
...
+#include <util/arith_tools.h>
+#include <util/c_types.h>
 #include <util/pointer_offset_size.h>
 #include <util/pointer_predicates.h>
 
-#include "utils.h"
+static const slicet normalize_to_slice(const exprt &expr, const namespacet &ns)
+{
+  // FIXME: Refactor these checks out to a common function that can be
+  // used both in compilation and instrumentation stages
```
Comment (feliperodri) [inline src/goto-instrument/contracts/assigns.cpp:26]: Are we tracking this request somewhere?

File: src/goto-instrument/contracts/assigns.cpp [hunk 24]
```diff
@@ -9,144 +9,177 @@ Date: July 2021
...
+  if(expr.id() == ID_pointer_object)
+  {
+    const auto &arg = expr.operands().front();
+    return {
+      minus_exprt{
+        typecast_exprt::conditional_cast(arg, pointer_type(char_type())),
+        pointer_offset(arg)},
```
Comment (feliperodri) [inline src/goto-instrument/contracts/assigns.cpp:33]: Shall we add a comment about why do we cast this to char here?

File: src/goto-instrument/contracts/assigns.cpp [hunk 25]
```diff
@@ -9,144 +9,177 @@ Date: July 2021
...
+    minus_exprt{plus_exprt{slice.first, slice.second},
+                from_integer(1, slice.second.type())},
+    location));
+
+  instructions.destructive_append(skip_program);
+  return instructions;
+}
+
+const exprt
+assigns_clauset::conditional_address_ranget::generate_unsafe_inclusion_check(
```
Comment (feliperodri) [inline src/goto-instrument/contracts/assigns.cpp:115]: Why unsafe?

File: src/goto-instrument/contracts/assigns.cpp [hunk 26]
```diff
@@ -9,144 +9,177 @@ Date: July 2021
...
+    return not_exprt{lhs.validity_condition_var};
 
-goto_programt
-assigns_clauset::generate_havoc_code(const source_locationt &location) const
-{
-  modifiest modifies;
+  exprt::operandst conditions{not_exprt{lhs.validity_condition_var}};
   for(const auto &target : write_set)
-    modifies.insert(target.address.object());
+    conditions.push_back(target.generate_unsafe_inclusion_check(lhs));
```
Comment (feliperodri) [inline src/goto-instrument/contracts/assigns.cpp:168]: So you call it `unsafe` because the validity checks are inserted separately?

File: src/goto-instrument/contracts/assigns.h (hunk 12 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/assigns.h:35]: Agreed.

File: src/goto-instrument/contracts/assigns.h [hunk 27]
```diff
@@ -16,59 +16,100 @@ Date: July 2021
...
+  exprt generate_inclusion_check(const conditional_address_ranget &) const;
+
+  const write_sett &get_write_set() const
+  {
+    return write_set;
+  }
 
   const messaget &log;
   const namespacet &ns;
+  const irep_idt &function_name;
```
Comment (feliperodri) [inline src/goto-instrument/contracts/assigns.h:92]: Why is this not `protected`?

File: src/goto-instrument/contracts/contracts.cpp [hunk 28]
```diff
@@ -673,29 +674,37 @@ void code_contractst::instrument_call_statement(
...
+      const auto &object = function_call.lhs();
+      // move past the call and then insert the CAR
+      instruction_it++;
+      const auto car = assigns.add_to_write_set(pointer_object(object));
+      auto snapshot_instructions = car->generate_snapshot_instructions();
+      insert_before_swap_and_advance(
+        body, instruction_it, snapshot_instructions);
+      // since CAR was inserted *after* the malloc,
+      // move the instruction pointer backward,
+      // because the caller increments it in a `for` loop
```
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.cpp:674]: Awesome!! Way clear now.

File: src/goto-instrument/contracts/utils.cpp [hunk 29]
```diff
@@ -10,6 +10,8 @@ Date: September 2021
 
 #include "utils.h"
 
+#include <util/c_types.h>
```
Comment (feliperodri) [inline src/goto-instrument/contracts/utils.cpp]: Why do we need this include?

File: regression/contracts/assigns_enforce_elvis_5/test.desc (hunk 7 above)
Comment (feliperodri) [reply regression/contracts/assigns_enforce_elvis_5/test.desc:12]: We should open a GitHub issue for this bug. This is outside the contract implementation.

**Review by feliperodri — COMMENTED** [review]
@SaswatPadhi only minor comments. Approved.

File: src/goto-instrument/contracts/assigns.cpp (hunk 25 above)
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/assigns.cpp:115]: Because validity is not checked for `lhs`.

File: src/goto-instrument/contracts/assigns.cpp (hunk 26 above)
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/assigns.cpp:168]: Yes exactly.

File: src/goto-instrument/contracts/assigns.h (hunk 27 above)
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/assigns.h:92]: It's `const` so it doesn't need to be

File: src/goto-instrument/contracts/utils.cpp (hunk 29 above)
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/utils.cpp]: Nice catch, we don't.

**Review by SaswatPadhi — APPROVED** [review]
Approving on behalf of @feliperodri.

File: regression/contracts/assigns_enforce_18/test.desc [hunk 30]
```diff
@@ -3,21 +3,18 @@ main.c
 --enforce-contract foo --enforce-contract bar --enforce-contract baz _ --pointer-primitive-check
 ^EXIT=10$
 ^SIGNAL=0$
-^\[bar.\d+\] line \d+ Check that \*a is assignable: SUCCESS$
+^\[bar.\d+\] line \d+ Check that POINTER_OBJECT\(\(void \*\)a\) is assignable: SUCCESS$
```
Comment (kroening) [inline regression/contracts/assigns_enforce_18/test.desc:6]: The new message is a downgrade in terms of user experience.

File: regression/contracts/assigns_enforce_23/main.c [hunk 31]
```diff
@@ -0,0 +1,41 @@
...
+{
+  uint8_t *buf;
+  size_t size;
+} blob;
+
+void foo(blob *b, uint8_t *value)
+  // clang-format off
+__CPROVER_requires(b->size == 5)
+__CPROVER_assigns(__CPROVER_POINTER_OBJECT(b->buf))
+__CPROVER_assigns(__CPROVER_POINTER_OBJECT(value))
```
Comment (kroening) [inline regression/contracts/assigns_enforce_23/main.c:14]: This overloaded use of `__CPROVER_POINTER_OBJECT` will confuse. I would advise to use a different identifier.

**Review by kroening — APPROVED** [review]

File: src/goto-instrument/contracts/utils.h [hunk 32]
```diff
@@ -53,10 +53,10 @@ class havoc_if_validt : public havoc_utilst
 /// over all dereferenced pointer expressions *(pexpr_1), *(pexpr_2), ...
 /// found in the AST of `expr`.
 ///
-/// \param expr The expression that contains dereferences to be validated
-/// \param ns The namespace that defines all symbols appearing in `expr`
+/// \param expr: The expression that contains dereferences to be validated.
```
Comment (chrisr-diffblue) [inline src/goto-instrument/contracts/utils.h:56]: Extra docs are always nice :-) But the addition of ':' following parameter names is unusual and adds a bit of noise to the diff. Not a blocking comment at all - just an observation to be aware of.

**Review by chrisr-diffblue — APPROVED** [review]

File: src/goto-instrument/contracts/utils.h (hunk 32 above)
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/utils.h:56]: I thought that's the convention in this repo (variable name followed by a colon in the docs). I saw that in several files, so changed this accordingly.
For example: https://github.com/diffblue/cbmc/blob/48893287099cb5780302fe9dc415eb6888354fd6/src/util/expr_cast.h
